node/lib/util/synthetic_gc_util.js

```diff
@@ -0,0 +1,330 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+"use strict";
+
+const assert  = require("chai").assert;
+const co = require("co");
+const NodeGit = require("nodegit");
+const SubmoduleUtil = require("./submodule_util");
+const SubmoduleFetcher = require("./submodule_fetcher");
+
+const SYNTHETIC_BRANCH_BASE = "refs/commits/";
+
+/**
+ * This class provides a way to list/delete synthetic refs, by following two
+ * main algorithms.
+ *  1) Removing refs that have persistent descendant.
+ *  2) Removing refs that are older than specific threshold date and not part
+ *     of the persistent descendant list.
+ *
+ * @SEE_ALSO: lib/synthetic_gc
+ */
+class SyntheticGcUtil {
+
+    /**
+    * Create a 'SyntheticGcUtil' object for synthetic refs manipulation.
+    */
+    constructor() {
+        this.d_visited = {};
+        this.d_simulation = true;
+    }
+
+    /**
+    * If simulation set to true, no synthetic refs are
+    * actually being removed.
+    * @param {Boolean}
+    */
+    get simulation() {
+        return this.d_simulation;
+    }
+
+    set simulation(value) {
+        this.d_simulation = value;
+    }
+
+    /**
+    * @param {Set} visited commits
+    */
+    get visited() {
+        return this.d_visited;
+    }
+
+    set visited(value) {
+        this.d_visited = value;
+    }
+
+} // SyntheticGcUtil
+
+/**
+ * Return bare submodule repository corresponding to a specified `refHeadCommit`
+ * within specified meta `repo`.
+ *
+ * @param {NodeGit.Repository}   repo
+ * @param {NodeGit.Commit}       commit
+ * @return {Number} 0 or an error code
+ */
+SyntheticGcUtil.prototype.getBareSubmoduleRepo = co.wrap(
+    function*(repo, subName, refHeadCommit) {
+
+        assert.instanceOf(repo, NodeGit.Repository);
+        assert.instanceOf(refHeadCommit, NodeGit.Commit);
+
+        const fetcher = new SubmoduleFetcher(repo, refHeadCommit);
+
+        const subUrl = yield fetcher.getSubmoduleUrl(subName);
+        const subRepo = yield NodeGit.Repository.open(subUrl);
+
+        return subRepo;
+});
+
+/**
+ * Remove synthetic ref corresponding to specified `commit` in the specified
+ * `repo`.
+ *
+ * @param {NodeGit.Repository}   repo
+ * @param {NodeGit.Commit}       commit
+ * @return {Number} 0 or an error code
+ */
+SyntheticGcUtil.prototype.removeSyntheticRef = co.wrap(
+    function(repo, commit) {
+
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(commit, NodeGit.Commit);
+
+    const synRefPath = SYNTHETIC_BRANCH_BASE + commit;
```

Comment:
I think we already had a function somewhere that does this?

Reply:
So, there this function in lib/synthetic_branch_util:60
`exports.getSyntheticBranchForCommit()`.

That I was calling before. But during the whole test ran it was very flaky by returning:
`return "refs/heads/metaTEST"`

I found a test that overrides this function in test/util/synthetic-branch.js:62
SyntheticBranch.getSyntheticBranchForCommit = function(commit) {                                                                 
            /*jshint unused:false*/                                                                                                      
         return "refs/heads/metaTEST";                                                                                                
};

I was not sure what is the proper way to fix it.
If you have any ideas I can submit pullrequest for it.

Reply:
I removed that test override, so you shouldn't have that problem anymore.